src/dbnode/encoding/series_iterator_accumulator.go

```diff
@@ -0,0 +1,256 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package encoding
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/ts"
+	"github.com/m3db/m3/src/x/ident"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+var _ SeriesIteratorAccumulator = (*seriesIteratorAccumulator)(nil)
+
+type seriesIteratorAccumulator struct {
+	id              ident.ID
+	nsID            ident.ID
+	start           time.Time
+	end             time.Time
+	iters           iterators
+	seriesIterators []SeriesIterator
+	err             error
+	firstNext       bool
+	closed          bool
+}
+
+// NewSeriesIteratorAccumulator creates a new series iterator.
+func NewSeriesIteratorAccumulator(
+	iter SeriesIterator,
+) (SeriesIteratorAccumulator, error) {
+	it := &seriesIteratorAccumulator{
+		id:              iter.ID(),
+		nsID:            iter.Namespace(),
+		seriesIterators: make([]SeriesIterator, 0, 2),
+	}
+
+	err := it.Add(iter)
+	if err != nil {
+		return nil, err
+	}
+
+	return it, nil
+}
+
+func (it *seriesIteratorAccumulator) Add(iter SeriesIterator) error {
+	if newNs := iter.Namespace(); !newNs.Equal(it.nsID) {
+		return fmt.Errorf("cannot add iterator with namespace %s to accumulator %s",
+			newNs.String(), it.nsID.String())
+	}
+
+	if !iter.Next() || !it.iters.push(iter) {
+		iter.Close()
+		return iter.Err()
+	}
+
+	iterStart := iter.Start()
+	if start := it.start; start.IsZero() || iterStart.Before(start) {
+		it.start = iterStart
+	}
+
+	iterEnd := iter.End()
+	if end := it.end; end.IsZero() || iterEnd.After(end) {
+		it.end = iterEnd
+	}
+
+	it.seriesIterators = append(it.seriesIterators, iter)
+	return nil
+
+}
+
+func (it *seriesIteratorAccumulator) ID() ident.ID {
+	return it.id
+}
+
+func (it *seriesIteratorAccumulator) Namespace() ident.ID {
+	return it.nsID
+}
+
+func (it *seriesIteratorAccumulator) Tags() ident.TagIterator {
+	it.err = errors.New("cannot get tags from series iterator accumulator")
```

Comment:
Maybe we should just avoid exposing "Tags()" on the "SeriesIteratorAccumulator" interface at all?

Reply:
If these methods are just going to return errors, I think it might be better to just declare a smaller interface that we can use and just update all places that currently reference a SeriesIterator with a SeriesIteratorAccmulator?